### DISCUSSION TITLE
Canonical way to pass environment variables to linters?

### DISCUSSION THREAD

**Discussion by dhermes** [golangci/golangci-lint#4253] [Q&A] [answered]
I'd like to set the `GOFUMPT_SPLIT_LONG_LINES=on` environment variable when running the `gofumpt` linter.

Is there a canonical way to do this with e.g. `.golangci.yaml`? Or should I just be setting it in the environment of the caller?

**Comment by Antonboom** [comment]
@dhermes hi

IMHO, canonical solution is 

1. Ask to add `no-long-lines` in [extra](https://github.com/mvdan/gofumpt/tree/master?tab=readme-ov-file#extra-rules-behind--extra) rules (or in separate option).
2. Use config
```yml
linters-settings:
  gofumpt:
    extra-rules: true
```
3. Use golangci-lint `--fix` option to autofix linter issues.

Or just use https://golangci-lint.run/usage/linters/#lll

**Comment by StevenACoffman** [comment]
An alternative is to not use gofumpt for splitting long lines at all, and instead use golines formatter, which is now maintained as one of the valid formatters for golangci-lint. 
- Related to #1914
- Related to [#5296 (comment)](https://github.com/golangci/golangci-lint/issues/5296#issuecomment-2577874799) 
- Related to #3983
- Related to #5432
- Related to #6657

**Comment by StevenACoffman** [comment, marked as answer]
According to https://github.com/mvdan/gofumpt/pull/358 gofumpt considers `GOFUMPT_SPLIT_LONG_LINES` an experimental feature that should not be exposed via `extra` or a separate flag.

In the interim, _do not_ set the `GOFUMPT_SPLIT_LONG_LINES` environment variable. 

`gofumpt` and `golines` have a _different_ default tab width (gofumpt:8, golines:4), and use _different_ split heuristics (gofumpt has minSplitFactor and per-context factors like 0.6 for top-level params, 1000 to suppress result splits; golines does dot-chain and argument wrapping differently). 

golangci-lint calls (if enabled) formatters in this order: gofmt → gofumpt → goimports → swaggo → gci → golines (last)

This configuration would _minimize_ the local environment's variable setting impact on the results, but they are only approximately equivalent:
```
# .golangci.yml
# All settings can be found here https://github.com/golangci/golangci-lint/blob/HEAD/.golangci.reference.yml
version: "2"
formatters:
  enable:
    - golines
    - gofumpt
  settings:
    golines:
      # NOTE: Avoid environment variable GOFUMPT_SPLIT_LONG_LINES=on
      # or note gofumpt/golines default tabWidth disagreement and
      # formatter resolution order:
      # gofmt → gofumpt → goimports → swaggo → gci → golines
      max-len: 100   # == default: 100, gofumpt's longLineLimit is 100
      tab-len: 8     # == default: 4, gofumpt's tabWidth is 8
      shorten-comments: false
```
